File: bindings/aster_python/aster/registry/acl.py
"""
aster.registry.acl — ACL enforcement for the Aster service registry.

Spec reference: Aster-SPEC.md §11.2.3.

ACL entries are stored in the registry doc at:
  _aster/acl/writers  → JSON list[str] of trusted writer AuthorIds
  _aster/acl/readers  → JSON list[str] of reader AuthorIds
  _aster/acl/admins   → JSON list[str] of admin AuthorIds

Phase 10 filters at READ TIME (post-read filter). Untrusted writes persist
locally but are excluded from all registry reads. True sync-time rejection
is deferred pending a future FFI hook.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING

from .keys import acl_key

if TYPE_CHECKING:
    pass  # DocHandle imported lazily

logger = logging.getLogger(__name__)


class RegistryACL:
    """In-memory ACL cache for the registry doc.

    Starts in *open mode* (all authors trusted) until an ACL entry is
    explicitly configured. Open mode is appropriate for local/dev deployments.

    When ``add_writer()`` is first called, the ACL switches to restricted mode
    and only explicitly listed writers are trusted.
    """

    def __init__(self, doc: object, author_id: str) -> None:
        """
        Args:
            doc: A ``DocHandle`` for the registry doc.
            author_id: The local author's ID (used to write ACL entries).
        """
        self._doc = doc
        self._author_id = author_id
        self._open = True               # All-trusted until a writer is set
        self._writers: set[str] = set()
        self._readers: set[str] = set()
        self._admins: set[str] = set()

    # ── Read-time filtering ───────────────────────────────────────────────────

    def is_trusted_writer(self, author_id: str) -> bool:
        """Return True if this author's writes should be trusted.

        Open mode: always True.
        Restricted mode: True only if author_id is in the writers set.
        """
        if self._open:
            return True
        return author_id in self._writers

    def filter_trusted(self, entries: list) -> list:
        """Filter a list of DocEntry objects to only trusted-author entries."""
        return [e for e in entries if self.is_trusted_writer(e.author_id)]

    # ── Reload from doc ────────────────────────────────────────────────────────

    async def reload(self) -> None:
        """Reload ACL state from the registry doc.

        If no ACL entries exist, stay in open mode. If entries are present,
        switch to restricted mode.
        """
        writers = await self._read_list("writers")
        if writers is None:
            # No ACL configured — remain open
            return
        readers = await self._read_list("readers") or []
        admins = await self._read_list("admins") or []
        self._writers = set(writers)
        self._readers = set(readers)
        self._admins = set(admins)
        self._open = False
        logger.debug(
            "ACL reloaded: %d writers, %d readers, %d admins",
            len(self._writers),
            len(self._readers),
            len(self._admins),
        )

    async def _read_list(self, subkey: str) -> list[str] | None:
        """Read a JSON list from the doc at ``_aster/acl/{subkey}``.

        Returns None if the key does not exist.
        """
        entries = await self._doc.query_key_exact(acl_key(subkey))
        # Prefer the entry written by ourselves (the admin)
        our_entries = [e for e in entries if e.author_id == self._author_id]
        target = our_entries[0] if our_entries else (entries[0] if entries else None)
        if target is None:
            return None
        raw = await self._doc.read_entry_content(target.content_hash)
        return json.loads(raw)

    # ── Admin operations ───────────────────────────────────────────────────────

    async def add_writer(self, author_id: str) -> None:
        """Add an author to the trusted writers set.

        Persists the updated list to the doc. Switches to restricted mode if
        this is the first explicit writer entry.
        """
        self._writers.add(author_id)
        self._open = False
        await self._persist("writers", list(self._writers))

    async def remove_writer(self, author_id: str) -> None:
        """Remove an author from the trusted writers set."""
        self._writers.discard(author_id)
        await self._persist("writers", list(self._writers))

    async def get_writers(self) -> list[str]:
        return list(self._writers)

    async def get_readers(self) -> list[str]:
        return list(self._readers)

    async def get_admins(self) -> list[str]:
        return list(self._admins)

    async def _persist(self, subkey: str, values: list[str]) -> None:
        await self._doc.set_bytes(
            self._author_id,
            acl_key(subkey),
            json.dumps(values).encode(),
        )

File: bindings/typescript/packages/aster/src/registry/acl.ts
/**
 * Registry ACL — access control list for registry doc entries.
 *
 * Mirrors bindings/python/aster/registry/acl.py.
 *
 * Starts in open mode (all authors trusted). When the first ACL
 * entry is set, switches to restricted mode.
 */

import { aclKey } from './keys.js';

/** Doc handle interface for ACL operations. */
interface AclDoc {
  setBytes(authorHex: string, key: string, value: Uint8Array): Promise<string>;
  getExact(authorHex: string, key: string): Promise<Uint8Array | null>;
}

/**
 * Registry ACL — filters doc entries by trusted authors.
 */
export class RegistryACL {
  private writers = new Set<string>();
  private readers = new Set<string>();
  private admins = new Set<string>();
  private _restricted = false;

  /** Whether ACL is in restricted mode. */
  get restricted(): boolean {
    return this._restricted;
  }

  /** Check if an author is a trusted writer. */
  isTrustedWriter(authorId: string): boolean {
    if (!this._restricted) return true;
    return this.writers.has(authorId) || this.admins.has(authorId);
  }

  /** Check if an author is a trusted reader. */
  isTrustedReader(authorId: string): boolean {
    if (!this._restricted) return true;
    return this.readers.has(authorId) || this.writers.has(authorId) || this.admins.has(authorId);
  }

  /**
   * Filter entries to only those from trusted authors.
   *
   * @param entries - Array of [authorId, value] pairs
   * @returns Filtered entries
   */
  filterTrusted<T extends { authorId: string }>(entries: T[]): T[] {
    if (!this._restricted) return entries;
    return entries.filter(e => this.isTrustedWriter(e.authorId));
  }

  /** Add a writer and switch to restricted mode. */
  async addWriter(authorId: string, doc?: AclDoc, adminAuthorId?: string): Promise<void> {
    this.writers.add(authorId);
    this._restricted = true;

    if (doc && adminAuthorId) {
      const encoder = new TextEncoder();
      const data = JSON.stringify([...this.writers]);
      await doc.setBytes(adminAuthorId, aclKey('writers'), encoder.encode(data));
    }
  }

  /** Add a reader. */
  addReader(authorId: string): void {
    this.readers.add(authorId);
    this._restricted = true;
  }

  /** Add an admin. */
  addAdmin(authorId: string): void {
    this.admins.add(authorId);
    this._restricted = true;
  }

  /** Remove a writer from the ACL. */
  removeWriter(authorId: string): void {
    this.writers.delete(authorId);
  }

  /** Get all writers. */
  getWriters(): string[] {
    return [...this.writers];
  }

  /** Get all readers. */
  getReaders(): string[] {
    return [...this.readers];
  }

  /** Get all admins. */
  getAdmins(): string[] {
    return [...this.admins];
  }

  /**
   * Reload ACL state from the registry doc.
   */
  async reload(doc: AclDoc, authorId: string): Promise<void> {
    const decoder = new TextDecoder();

    const writersData = await doc.getExact(authorId, aclKey('writers'));
    if (writersData) {
      const list: string[] = JSON.parse(decoder.decode(writersData));
      this.writers = new Set(list);
      this._restricted = true;
    }

    const readersData = await doc.getExact(authorId, aclKey('readers'));
    if (readersData) {
      const list: string[] = JSON.parse(decoder.decode(readersData));
      this.readers = new Set(list);
    }

    const adminsData = await doc.getExact(authorId, aclKey('admins'));
    if (adminsData) {
      const list: string[] = JSON.parse(decoder.decode(adminsData));
      this.admins = new Set(list);
    }
  }
}
